// meta-test/test/plugins.node.js
var helper = require('./lib/helper')
  , it = require('it-is')
  , Plugins = require('../plugins')

exports ['plugins calls default'] = function (){
  var plugins = 
    new Plugins ({hook: function default1(){return 10}})
  it(plugins).property('hook',it.function())
  
  it(plugins.hook()).equal(10)
}

exports ['plugins calls added module first'] = function (){
  var incrementor = 
    { hook: function plugged(fallback){
        return fallback() + 1
      }
    }

  var plugins = 
    new Plugins ({hook: function default2(){return 10}})
    .add(incrementor)

  it(plugins).property('hook',it.function())
  it(plugins.hook()).equal(11)
  
  plugins 
    .add(incrementor)

  it(plugins.hook()).equal(12)

  plugins 
    .add(incrementor)

  it(plugins.hook()).equal(13)
}

exports ['3 let absent hooks fall through'] = function (){

  var plug1 = 
      { a: function doubleVal (val,def){
          return def(0) + (val * 2)
        }
      }
    , plug2 = {
        b: function sayHelloThere(name,def){
          re
